src/lib.rs

```diff
@@ -505,180 +248,26 @@ pub fn encode_without_checksum<T: AsRef<[u5]>>(hrp: Hrp, data: T) -> Result<Stri
     Ok(buf)
 }
 
-/// Decodes a bech32 string into the raw HRP and the data bytes.
-///
-/// # Returns
-///
-/// The human-readable part in lowercase, the data with the checksum removed, and the encoding.
-#[cfg(feature = "alloc")]
-pub fn decode(s: &str) -> Result<(Hrp, Vec<u5>, Variant), Error> {
-    let (hrp_lower, mut data) = split_and_decode(s)?;
-    if data.len() < CHECKSUM_LENGTH {
-        return Err(Error::InvalidLength);
+/// Decodes a bech32 string.
+pub fn decode(s: &str) -> Result<(hrpstring::Parsed, Variant), Error> {
+    if let Ok(p) = Parsed::new::<Bech32m>(s) {
+        return Ok((p, Variant::Bech32m));
     }
-
-    // Ensure checksum
-    match verify_checksum(hrp_lower, &data) {
-        Some(variant) => {
-            // Remove checksum from data payload
-            data.truncate(data.len() - CHECKSUM_LENGTH);
-
-            Ok((hrp_lower, data, variant))
-        }
-        None => Err(Error::InvalidChecksum),
+    match Parsed::new::<Bech32>(s) {
+        Ok(p) => Ok((p, Variant::Bech32)),
+        Err(e) => Err(Error::Hrpstring(e)),
```

Comment:
In e2527dc19c3e8e9a4cd678a955fc5b161e175c87:

Is there any way with this API to avoid verifying the entire checksum twice?

Reply:
That's an optimization, right? I didn't put too much thought into how the crate API interacts with the new `primitives` code, except to ensure that we do not take any functionality away. I think we should tackle these questions when we re-write the whole API and/or add a segwit one.

Reply:
Oh by "re-write the whole API" I meant the public functions in `lib.rs` and if we add a segwit specific API (along with all the test vectors for bip 173/350), I have all this started in other branches from the last few months.